osquery/events/linux/auditeventpublisher.cpp

```diff
@@ -122,15 +128,57 @@ void AuditEventPublisher::ProcessEvents(
 
       event_context->audit_events.push_back(audit_event);
 
-      // SELinux events
+      // SELinux or AppArmor events
     } else if (selinux_event_set.find(audit_event_record.type) !=
                selinux_event_set.end()) {
-      AuditEvent audit_event;
-      audit_event.type = AuditEvent::Type::SELinux;
-      audit_event.record_list.push_back(audit_event_record);
-
-      event_context->audit_events.push_back(audit_event);
-
+      if (audit_event_record.fields.find(kAppArmorEventMarker) ==
```

Comment:
Oh wow, this whole function, `ProcessEvents`, is way too complex. We should refactor this in the near future.

Reply:
I am worried that we are doing searches in performance-critical code. Is there a way to speed up this search, for example, will the marker always appear first?

Reply:
Actually it's a search in a map (check out `AuditEventRecord` declaration), so it should be fine as it takes `O(log n)`. The only speed up suggestion for this place is the use of `unordered_map` which has constant complexity for searches by key. IMHO in this case it's not that critical as the size of this map is usually small.

Reply:
> Oh wow, this whole function, `ProcessEvents`, is way too complex. We should refactor this in the near future.

I'm ready to help, but lets do it in a separate PR.